nab/detectors/htmjava/htmjava_detector.py

```diff
@@ -0,0 +1,130 @@
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import math
+import simplejson as json
+from subprocess import Popen, PIPE
+
+from nupic.algorithms import anomaly_likelihood
+from nupic.frameworks.opf.common_models.cluster_params import (
+  getScalarMetricWithTimeOfDayAnomalyParams)
+
+from nab.detectors.base import AnomalyDetector
+
+
+class HtmjavaDetector(AnomalyDetector):
+  """
+  Inspired by the 'NumentaDetector' replacing the 'OPF CLAModel' with a
+  java subprocess running 'htm.java' model
+  """
+  def __init__(self, *args, **kwargs):
+
+    super(HtmjavaDetector, self).__init__(*args, **kwargs)
+
+    self.model = None
+    self.sensorParams = None
+    self.modelParams = None
+    self.anomalyLikelihood = None
+
+
+  def getAdditionalHeaders(self):
+    """Returns a list of strings."""
+    return ["raw_score"]
+
+
+  def handleRecord(self, inputData):
+    """Returns a tuple (anomalyScore, rawScore).
+
+    Internally to NuPIC "anomalyScore" corresponds to "likelihood_score"
+    and "rawScore" corresponds to "anomaly_score". Sorry about that.
+    """
+    # Send input to HTM Java detector
+    line = "{0},{1}\n".format(inputData['timestamp'], inputData['value'])
+    self.model.stdin.writelines(line)
+
+    # Retrieve the anomaly score
+    result = self.model.stdout.readline()
+    rawScore = float(result)
+
+    # Compute log(anomaly likelihood)
+    anomalyScore = self.anomalyLikelihood.anomalyProbability(
+      inputData["value"], rawScore, inputData["timestamp"])
+    logScore = self.anomalyLikelihood.computeLogLikelihood(anomalyScore)
+
+    return (logScore, rawScore)
+
+
+  def initialize(self):
+    # Get config params, setting the RDSE resolution
+    rangePadding = abs(self.inputMax - self.inputMin) * 0.2
+    self.modelParams = getScalarMetricWithTimeOfDayAnomalyParams(
+      metricData=[0],
+      minVal=self.inputMin-rangePadding,
+      maxVal=self.inputMax+rangePadding,
+      minResolution=0.001
+    )["modelConfig"]
+
+    self._setupEncoderParams(
+      self.modelParams["modelParams"]["sensorParams"]["encoders"])
+
+    # Initialize the anomaly likelihood object
+    numentaLearningPeriod = math.floor(self.probationaryPeriod / 2.0)
+    self.anomalyLikelihood = anomaly_likelihood.AnomalyLikelihood(
+      claLearningPeriod=numentaLearningPeriod,
+      estimationSamples=self.probationaryPeriod-numentaLearningPeriod,
+      reestimationPeriod=100
+    )
+
+    if self.model:
```

Comment:
What is this for? I'm assuming to close the pipe.
It would be better to have a `_self.terminateJavaPipe()`method that does this, and is also called at the end of `run()`.